### DISCUSSION TITLE
What targets are supported?

### DISCUSSION THREAD

**Discussion by marcpabst** [juntyr/pyodide-wasm-runtime-layer#2] [Q&A] [answered]
Hi, this looks impressive!

This is more of a question for clarification than an actual issue, but since discussions aren't enabled, I hope it's okay to open an issue instead. I'm trying to understand how to use this layer. Can I compile to one of the `wasi` targets and then run my code in the browser? Or should I target `wasm32-unknown-unknown`? I'm not very familiar with the concept of WASI runtime layers and would appreciate some help!

Thank you!

**Comment by juntyr** [comment, marked as answer]
Thank you for your question!

The short answer is that both the wasm32-unknown-unknown and wasm32-wasi targets will work with the runtime layer, but that you’ll have an easier time getting started with the former.

The wasm-runtime-layer crate and its backends, including this crate, provide an abstraction over core WASM runtimes where you have to do all linking explicitly. If you compile your WASM module code to the wasm32-unknown-unknown target, you have to explicitly define every interaction surface with the host, and so getting the linking done is quite straightforward. If you compile to wasm32-wasi, you will need to link in a host implementation of the WASI interface (including file system, stdio, clock, network support etc). While WASI support is made easy with e.g. wasmtime and support on the web is slowly coming e.g. with jco, the wasm-runtime-layer crate doesn’t have an easy drop-in WASI implementation yet.

If you are using the WASM component model and WASI preview 2, you will likely want to use the wasm-component-layer crate, which is built on wasm-runtime-layer and allows you to run WASM components e.g. on the web using this pyodide-webassembly-runtime-layer crate.

(I have an unpublished implementation of WASI preview 2 that I’m working on, which doesn’t actually expose anything but instead sandboxes the WASM component while allowing it to think that it has WASI access, similar to what the wasm-virt project does. If that’s interesting, I can try to publish it a bit sooner.)

Finally, of course, I have to note that this crate is for a very niche use case, namely when you want to write a Python module in Rust, using PyO3, that should run on the browser inside Pyodide, i.e. inside WASM, but also needs to run WASM modules itself. What you end up with is running WASM modules from within a Python runtime in WASM (which is quite cool).

If you need the more common use case of just executing WASM code on the web by coupling your Rust code to JS, the js-wasm-runtime-layer backend will work best for you.

I hope this answer is helpful - please feel free to ask for clarifications :) If you tell me more about your use case, I can also try to share some of my own code snippets to help you get started.
